tunneld/tunneld_test.go

```diff
@@ -0,0 +1,222 @@
+package tunneld_test
+
+import (
+	"context"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"reflect"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/wgtunnel/tunneld"
+	"github.com/coder/wgtunnel/tunnelsdk"
+)
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+
+	td, client := createTestTunneld(t, nil)
+	require.NotNil(t, td)
+	require.NotNil(t, client)
+
+	res, err := client.Request(context.Background(), http.MethodGet, "/", nil)
+	require.NoError(t, err)
+	_ = res.Body.Close()
+	require.Equal(t, http.StatusNotFound, res.StatusCode)
+}
+
+// TestEndToEnd does an end-to-end tunnel test by creating a tunneld server, a
+// client, setting up the tunnel, and then doing a bunch of tests through the
+// tunnel to ensure it works.
+func TestEndToEnd(t *testing.T) {
+	t.Parallel()
+
+	td, client := createTestTunneld(t, nil)
+	require.NotNil(t, td)
+
+	// Start a tunnel.
+	key, err := tunnelsdk.GeneratePrivateKey()
+	require.NoError(t, err, "generate private key")
+	tunnel, err := client.LaunchTunnel(context.Background(), tunnelsdk.TunnelConfig{
+		Log: slogtest.
+			Make(t, &slogtest.Options{IgnoreErrors: true}).
+			Named("tunnel_client"),
+		PrivateKey: key,
+	})
+	require.NoError(t, err, "launch tunnel")
+	t.Cleanup(func() {
+		_ = tunnel.Close()
+		<-tunnel.Wait()
+	})
```

Comment:
Personally I still think defers make sense for this. `t.Cleanup` should be used for when stuff should be cleaned up outside of the current function.